src/main/java/com/example/SOPSbackend/service/PatientService.java

```diff
@@ -2,31 +2,84 @@
 
+import java.util.Collection;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
 @Service
 public class PatientService {
     private final PatientRepository patientRepository;
     private final PasswordEncoder passwordEncoder;
+    private final VaccineRepository vaccineRepository;
+    private final VaccinationSlotRepository vaccinationSlotRepository;
+    private final VaccinationRepository vaccinationRepository;
 
-    public PatientService(PatientRepository patientRepository, PasswordEncoder passwordEncoder) {
+    public PatientService(PatientRepository patientRepository, PasswordEncoder passwordEncoder,
+                          VaccineRepository vaccineRepository, VaccinationSlotRepository vaccinationSlotRepository,
+                          VaccinationRepository vaccinationRepository) {
         this.patientRepository = patientRepository;
         this.passwordEncoder = passwordEncoder;
+        this.vaccineRepository = vaccineRepository;
+        this.vaccinationSlotRepository = vaccinationSlotRepository;
+        this.vaccinationRepository = vaccinationRepository;
     }
 
     public PatientEntity register(NewPatientRegistrationDto registration) throws UserAlreadyExistException {
-        if(checkIfUserExistByPesel(registration.getPesel()))
+        if (checkIfUserExistByPesel(registration.getPesel()))
             throw new UserAlreadyExistException("User already exists for this pesel");
-        if(checkIfUserExistByEmail(registration.getEmail()))
+        if (checkIfUserExistByEmail(registration.getEmail()))
             throw new UserAlreadyExistException("User already exists for this email");
 
         return patientRepository.save(new PatientEntity(registration, passwordEncoder));
     }
 
+    public List<VaccineEntity> getVaccines(Collection<String> diseases) {
+        return vaccineRepository.findByDiseaseIn(diseases);
+    }
+
+    public List<VaccinationSlotEntity> getAvailableVaccinationSlots() {
+        return vaccinationSlotRepository.findAvailableSlots();
+    }
+
+    public void reserveVaccinationSlot(long vaccineId, long vaccinationSlotId, PatientEntity patient) throws AlreadyReservedException {
+        Optional<VaccineEntity> vaccine = vaccineRepository.findById(vaccineId);
+        Optional<VaccinationSlotEntity> vaccinationSlot = vaccinationSlotRepository.findById(vaccinationSlotId);
+
+        if (!vaccine.isPresent()) {
+            throw new NoSuchElementException("Vaccine not found");
+        }
+
+        if (!vaccinationSlot.isPresent()) {
+            throw new NoSuchElementException("Vaccination slot not found");
+        }
+
+        VaccinationEntity vaccination = vaccinationRepository.findByVaccinationSlot(vaccinationSlot.get());
+
+        if (vaccination != null) {
+            throw new AlreadyReservedException("Vaccination slot is already reserved.");
+        }
+
+        vaccinationRepository.save(new VaccinationEntity(
```

Comment:
Podoba mi się tutaj bardzo ta technika "guardów" i wykorzystanie wyjątków (chociaż źle zdefiniowanych). Widać jak powoli dochodzimy do tego co się ma stać i jak dużo założeń musi być spełnionych przed otrzymaniem nagrody jaką jest wykonanie linijek 76-80. **Pytanie czy niektórych z tych założeń nie da się wyeliminować w jakiś lepszy sposób.**

IMO

Reply:
Wygląda na to, że się nie da (to naprawde bardzo przemyślany kod 😍).